<!DOCTYPE html>
<html>
<head>
<title>PRANK</title>
<link rel="stylesheet" type="text/css" href="docs/style.css" />
</head>
<body>
<h1 id="prank">PRANK</h1>
<p>PRANK is a probabilistic multiple alignment program for DNA, codon and amino-acid sequences. It’s based on a novel algorithm that treats insertions correctly and avoids over-estimation of the number of deletion events. In addition, PRANK borrows ideas from maximum likelihood methods used in phylogenetics and correctly takes into account the evolutionary distances between sequences. Lastly, PRANK allows for defining a potential structure for sequences to be aligned and then, simultaneously with the alignment, predicts the locations of structural units in the sequences.</p>
<hr>
<p><img src="docs/data/prank_logo.png" alt=""></p>
<ul>
<li><a href="https://github.com/ariloytynoja/prank-msa/tree/master/binaries">PRANK binaries</a></li>
</ul>
<p>Packed source code and pre-compiled binaries for Linux, Windows and OSX.</p>
<ul>
<li><a href="#using-prank">Using PRANK</a></li>
</ul>
<p>Instructions for the use of PRANK with example commands and test data.</p>
<ul>
<li><a href="docs/prank_installation.html">Installing PRANK</a></li>
</ul>
<p>Instructions to install and start using PRANK on different operating systems. Pre-compiled executables are provided for Windows, Mac OSX and Linux; the source code should compile on any system supporting the GNU C compiler package.</p>
<ul>
<li><a href="docs/prank_new_features.html">New features</a></li>
</ul>
<p>Brief description of the new features in the latest versions of the program.</p>
<ul>
<li><a href="docs/prank_known_issues.html">Known issues</a></li>
</ul>
<p>Known bugs and problems with the latest versions of the program.</p>
<ul>
<li><a href="docs/prank_structure_models.html">Structure models</a></li>
</ul>
<p>Generation of PRANK structure models.</p>
<ul>
<li><a href="docs/hsaml_format.html">HSAML format</a></li>
</ul>
<p>Definition of the XML-format used by PRANK.</p>
<hr>
<h3 id="using-prank">Using PRANK</h3>
<p>At the simplest, PRANK can be run with command:  
 </p>
<pre><code>prank input_file
</code></pre>
<p>where <code>input_file</code> contains sequences in FASTA format.</p>
<p>See below for the description of the most central program options.</p>
<ul>
<li><a href="#using-prank">Using PRANK</a></li>
<li><a href="#about-prank-alignments">About PRANK alignments</a></li>
<li><a href="#main-program-options">Main program options</a></li>
<li><a href="#other-program-options">Other program options</a><ul>
<li><a href="#translated-alignment-and-codon-alignment">Translated alignment and codon alignment</a></li>
<li><a href="#inference-of-ancestral-sequences-and-ancestral-events">Inference of ancestral sequences and ancestral events</a></li>
<li><a href="#finishing-an-alignment-and-merging-two-alignments">Finishing an alignment and merging two alignments</a></li>
<li><a href="#alignment-reformatting-and-back-translation">Alignment reformatting and back-translation</a></li>
</ul>
</li>
<li><a href="#methods">Methods</a></li>
</ul>
<h3 id="about-prank-alignments">About PRANK alignments</h3>
<p>PRANK aims at an evolutionarily correct alignment and the alignments inferred with PRANK can be expected to look different from ones generated with other alignment methods. There are, however, cases where the different look is caused by violations of the method’s assumptions. To understand why things may go wrong and how to avoid that, read this <a href="docs/prank_differences.html">explanation of differences</a> between PRANK and traditional progressive alignment methods.</p>
<p>The reconstruction of evolutionary homology — including the correct placement of insertion and deletion events — is only feasible for rather closely-related sequences. PRANK is not meant for the alignment of very diverged sequences. If sequences are very different, the correct homology cannot be reconstructed with confidence and PRANK may simply refuse to match them. There are several methods developed for structural matching of very distant protein sequences. One should not consider the resulting alignment a proper inference of evolutionary homology, though.</p>
<p>Often there are ties in the alignment, i.e. positions with many equally good solutions. The common practice among sequence aligners is always to pick the same solution among many different ones and produce consistent alignments. We believe that this gives the user a wrong impression and too high confidence on the resulting alignment (for example, 10 iterations of the alignment always produce the same result -&gt; the alignment must be correct). Our decision is to break the ties randomly. This means that different runs of the program with the same data may give different alignments and PRANK alignment may not be reproducible. However, the practice we have chosen also tells the user that the regions not consistently aligned in a similar manner simply cannot be resolved reliably. Be aware that the reproducability in many other methods does not mean higher confidence — they also reproduce exactly the same errors!</p>
<h3 id="main-program-options">Main program options</h3>
<p>A typical command for PRANK could be:</p>
<pre><code>prank -d=input_file -t=tree_file -o=output_file -F -showxml
</code></pre>
<p>Here, <code>input_file</code> is the name of the file with input sequences in FASTA format, <code>tree_file</code> is the name of the file with a phylogeny with branch lengths relating these sequences, and <code>output_file</code> is the name of the file (with extension .best.fas) where the resulting alignment will be written in FASTA format; <code>-F</code> specifies that the inference of insertions should be trusted and sites appearing as insertions should not be aligned at the later stages of the process; and <code>-showxml</code> specifies that the output is also written in HSAML format, compatible for Wasabi import (including the alignment phylogeny).</p>
<p>Of these parameters, only <code>-d=input_file</code> is strictly required (and, when given alone, can be given in the form <code>prank input_file</code>). If the option <code>-t=tree_file</code> is omitted, an alignment guide tree is inferred from the input data (see below). If the option <code>-o=output_file</code> is omitted, the results are written to file named output.best.fas. Finally, if option <code>-F</code> is omitted, the algorithm does not incorrectly penalise for gaps for insertions but it will also not guarantee to leave the insertions unmatched.  
 </p>
<p>By default, PRANK iterates the alignment five times, inferring a new guide tree after each alignment. For each solution, it computes a phylogeny-aware alignments score and, after the full analysis, outputs the best-scoring solution. The number of iterations performed can be changed with option <code>-iterate=#</code> and the iteration completely disabled with option <code>-once</code>. See <a href="#methods">Methods</a> for more details.  
 </p>
<p>Without additional options, PRANK outputs only the alignment. More information can be outputted with the following options:  
 </p>
<pre><code>-showtree: output the guide tree used  
-showxml: output the results in PRANK’s own XML format  
-showanc: output ancestral sequences and guide tree  
-showevents: output events per branch  
-showall: output all listed above  
-showiter: output results for each iteration  
</code></pre>
<p>If a guide tree is provided, PRANK assumes that the sequence names in the tree and data file match exactly. Special characters and white spaces in the names may cause problems and may have to be removed/replaced with underscores. With option <code>-shortnames</code> only the first part (until the first white space) of the names in the data file are matched against the names in the guide tree. Extra names in the guide tree/data file can be removed with options <code>-prunetree</code> and <code>-prunedata</code>.  
 </p>
<p>A list of possible program options is shown with command:  
 </p>
<pre><code>prank -help
</code></pre>
<p>By default, PRANK writes the output alignments in FASTA format. With the option -f=format_name some other popular formats can be selected. See prank -help for details.</p>
<p><a href="#using-prank">back to top</a></p>
<h3 id="other-program-options">Other program options</h3>
<h4 id="translated-alignment-and-codon-alignment">Translated alignment and codon alignment</h4>
<p>PRANK can do translated alignments of protein-coding DNA sequences or align them using the codon model. Translation is selected with the options <code>-translate</code> (standard code) or <code>-mttranslate</code> (mitochondrial code), and the codon alignment with the option <code>-codon</code>. Using the example data <a href="docs/data/input_dna.fas">input_dna.fas</a>, the following commands make a translated alignment:  
 </p>
<pre><code>prank -d=input_dna.fas -o=output_translated -translate -F
</code></pre>
<p>and a codon alignment:  
 </p>
<pre><code>prank -d=input_dna.fas -o=output_codon -codon -F
</code></pre>
<p> <br>See <a href="#methods">Methods</a> for more details.</p>
<h4 id="inference-of-ancestral-sequences-and-ancestral-events">Inference of ancestral sequences and ancestral events</h4>
<p>The key aim of the phylogeny-aware alignment algorithm is to correctly model the insertion and deletion events and accurately infer the ancestral sequences, especially the presence and absence of characters. The ancestral sequences, that PRANK internally will anyway infer, can be of great interest and can also be outputted. The earlier output format for the ancestral sequences was not clear and new option <code>-showanc</code> has been added. This option now inserts the ancestral sequences in the position they appear in the alignment tree structure and, when viewed together with a phylogenetic tree with internal nodes labelled accordingly, make it easy to assign evolutionary events to specific tree branches. The two output files are named <code>*.best.anc.fas</code> and <code>*.best.anc.dnd</code>; with the option <code>-showevents</code>, the inferred events are listed per branch and outputted to file <code>*.best.events</code>.  
 </p>
<p>Ancestral sequences can be outputted for alignments generated with PRANK but they can also be inferred for existing alignments. While PRANK by default removes all the gap characters in the input data, this can be disabled with option <code>-keep</code> and it then reads the alignment as it is. Thus with command:  
 </p>
<pre><code>prank -d=alignment_pep.fas -showanc -showevents -keep -o=output_ancestors -njtree [OR -t=treefile]
</code></pre>
<p>PRANK will infer the ancestral sequences for an exiting alignment (this one available at <a href="docs/data/alignment_pep.fas">alignment_pep.fas</a>).  
 </p>
<p>If the aim is infer ancestral codon sequences but the codon alignment itself is found too slow, one can make the alignment using the protein model and then infer the ancestors for the back-translated DNA alignment:  
 </p>
<pre><code>prank -d=input.fas -o=output_translate -showall -translate
</code></pre>
<p>The ancestral sequences will be outputted to <code>output_translate_codon.[nuc|pep\].anc.fas</code>.  
 </p>
<p>See <a href="#methods">Methods</a> for more details.</p>
<h4 id="finishing-an-alignment-and-merging-two-alignments">Finishing an alignment and merging two alignments</h4>
<p>PRANK has new features that allow both aligning two alignments and finishing an alignment of a partially resolved dataset, consisting of one of more aligned sub-trees.  
 </p>
<p>In the first case, the two alignments and the corresponding trees (for example, <a href="docs/data/alignment_1.fas">alignment_1.fas</a>, <a href="docs/data/alignment_2.fas">alignment_2.fas</a>, <a href="docs/data/tree_1.tre">tree_1.tre</a> and <a href="docs/data/tree_2.tre">tree_2.tre</a>) are defined as:  
 </p>
<pre>prank -d1=alignment_1.fas -d2=alignment_2.fas -t1=tree_1.tre -t2=tree_2.tre -tree=&quot;(t1:0.1,t2:0.2);&quot;
</pre>
<p>If no trees (<code>-t1=</code> and <code>-t2=</code>) are provided, PRANK will compute NJ trees for the two alignments and infer ancestors based on those. If no merge-tree (<code>-t=</code> or <code>-tree=</code>) is provided, the default branch length is used; that can be modified with option <code>-mergedist=</code>. The merging of the two alignments is based on the alignment of the two ancestors.  
 </p>
<p>In the second case, the sequences belonging to the same group are indicated by ” group_XX” in the end of the name (where XX is unique for each group) and PRANK automatically finds the maximally large sub-trees consisting of sequences from the same group.  
 </p>
<p>This feature is best explained by an example. If you download files <a href="docs/data/example_partaligned.fas">example_partaligned.fas</a> and <a href="docs/data/example.tre">example.tre</a>, the alignment can be finished with the command:  
 </p>
<pre><code>prank -d=example_partaligned.fas -t=example.tre -partaligned
</code></pre>
<p>When finishing an alignment, PRANK looks for maximally large sub-trees of pre-aligned sequences belonging to the same group. If a sub-tree includes a sequence not from the same group, that sequence is normally aligned and also sequences outside that sequence (even if they belong to that group) will be re-aligned. This is again better explained by an example. If you download files <a href="docs/data/example_partaligned_2.fas">example_partaligned_2.fas</a> and run the command:  
 </p>
<pre><code>prank -d=example_partaligned_2.fas -t=example.tre -partaligned
</code></pre>
<p>you notice that many more nodes need to be aligned in order to finish the complete alignment. Note that PRANK will fail if the group annotation does not match actual data and sequences assigned to a specific group are not truly aligned.</p>
<h4 id="alignment-reformatting-and-back-translation">Alignment reformatting and back-translation</h4>
<p>PRANK supports several different alignment formats and can translate and back-translate sequence data between DNA and protein. These features can be exploited also without performing alignment of sequences. Example data used below can be found at <a href="docs/data/alignment_codon.fas">alignment_codon.fas</a>, <a href="docs/data/alignment_pep.fas">alignment_pep.fas</a> and <a href="docs/data/input_dna.fas">input_dna.fas</a>.  
 </p>
<p>Option <code>-convert</code> tells to convert between formats. This is typically accompanied with option <code>-f=format</code> where format is either <code>fasta</code>, <code>phylipi</code>, <code>phylips</code>, <code>paml</code>, <code>nexus</code> or <code>raxml</code>. The input and output files are specified as normally with options <code>-d=filename</code> and <code>-o=filename</code>; with options <code>-translate</code> and <code>-mttranslate</code> the DNA sequences are additionally translated to proteins. The command:  
 </p>
<pre><code>prank -convert -d=alignment_codon.fas -translate -f=phylipi -o=alignment_pep -keep
</code></pre>
<p>converts a codon alignment in FASTA format to a protein alignment in interleaved PHYLIP format.  
 </p>
<p>If option <code>-dna=filename</code> is included, PRANK attempts to back-translate the input protein alignment to the corresponding DNA alignment. This assumes that the input alignment (<code>-d=filename</code>) is proteins and the the other file (<code>-dna=filename</code>) contains matching unaligned DNA sequences with identical names. The command:  
 </p>
<pre><code>prank -convert -d=alignment_pep.fas -dna=input_dna.fas -o=alignment_dna -keep
</code></pre>
<p>converts a protein alignment in FASTA format to a DNA alignment in FASTA format.</p>
<p><a href="#using-prank">back to top</a></p>
<h2 id="methods">Methods</h2>
<h4 id="models">Models</h4>
<p>For DNA data, PRANK by default uses HKY model with empirical base frequencies and kappa=2. With the optional command parameters, it supports TN <a href="http://mbe.oxfordjournals.org/content/10/3/512.short">(Tamura and Nei, 1993)</a> and models below it (JC, K2P, FEL, HKY). For example, JC model is defined as <code>-kappa=1 -dnafreqs=0.25,0.25,0.25,0.25</code>. WAG <a href="http://mbe.oxfordjournals.org/content/18/5/691.short">(Whelan and Goldman, 2001)</a> is used for protein alignments.  
 </p>
<p>PRANK can do translated alignments of protein-coding DNA. It translates protein-coding DNA to protein sequences, aligns these sequences as proteins, and back-translates the resulting alignment to DNA such that the gaps are maintained. (PRANK can also back-translate protein alignments produced with external alignment software; see below for examples.)  
 </p>
<p>In addition to translated alignment, PRANK can also align codon sequences using a codon substitution matrix <a href="http://mbe.oxfordjournals.org/content/24/7/1464.short">(Kosiol, Holmes and Goldman, 2007)</a>. Translation into amino acids and codons is done in the first forward frame without any error-checking. Based on independent benchmarks, codon alignment produces more accurate alignments than alignment of translated protein sequences.  
 </p>
<p>Simulation studies with nucleotide sequences containing high numbers of insertions and deletions showed that the option <code>-F</code> gives the most accurate results and should be used when the guide phylogeny can be trusted <a href="http://www.sciencemag.org/content/320/5883/1632.short">(Löytynoja and Goldman, 2008)</a>.  
 </p>
<h4 id="guide-tree">Guide tree</h4>
<p>Progressive alignment requires a guide tree, and the algorithm of PRANK can be especially sensitive to errors in that. If a trusted pre-existing phylogeny with branch lengths is available for the input sequences (as often is in comparative genomic studies), it is recommended to use that.  
 </p>
<p>If no tree is provided, PRANK constructs one: it first call MAFFT <a href="http://nar.oxfordjournals.org/content/33/2/511.short">(Katoh et al, 2005)</a> to make a quick alignment and infers an NJ tree from the evolutionary distances based on that; an alternative (and much slower) approach is to estimate the evolutionary distances from pairwise alignments generated by PRANK itself.  
 </p>
<p>By default, PRANK iterates the alignment five times and writes the solution with the best score to the file named filename.best.fas. To prevent alignment iteration (e.g. when providing a guide tree that you trust), use the flag <code>-once</code> or <code>-iterate=1</code>.  
 </p>
<h4 id="anchoring">Anchoring</h4>
<p>The standard PRANK algorithm is based on an exhaustive search of the best pairwise solution, and for long sequences this soon becomes too time consuming. By default PRANK uses Exonerate <a href="http://www.biomedcentral.com/1471-2105/6/31">(Slater and Birney, 2005)</a> to anchor the pairwise alignments and thus speed up the process.  
 </p>
<h4 id="ancestral-reconstruction">Ancestral reconstruction</h4>
<p>PRANK’s own ancestral reconstruction happens during progressive alignment and is based on descendant sequences only. To improve the ancestral reconstruction, PRANK uses BppAncestor from the BppSuite package <a href="http://www.biomedcentral.com/1471-2148/8/255">(Dutheil and Boussau, 2008)</a> to infer the character states and then combines these with the inferred character presence/absence information. PRANK’s optimization score is also based on these maximum likelihood inferences.</p>
<p><a href="#using-prank">back to top</a></p>
</body>
</html>