<h1 id="hsaml-format">HSAML format</h1>
<p><a href="../index.html">Back to PRANK home.</a></p>
<p>Alignments generated with PRANK contain more information than can be stored using the traditional sequence alignment formats, so we have defined a new alignment format called HSAML. Our new web-tool, <a href="http://wasabiapp.org/" title="Wasabiapp">Wasabi</a>, uses this format and provides a superb graphical front-end to open and browse HSAML alignments.</p>
<p>The HSAML format follows the XML standards and, in addition to Wasabi, any program supporting the XML is able to open and manipulate the alignments. Two of the possible software packages to handle the data are R (see <a href="http://www.r-project.org/">here</a>), a free software environment for statistical computing and graphics, and perl (see <a href="http://www.perl.org/">here</a>), a cross-platform programming language. We will not go into details of R and perl here but provide simple examples how to manipulate HSAML-formatted data using these two packages. Which ever is your choice of a programming language, it is strongly recommended that you use one of the existing XML parsers to import the HSAML alignments and will not try to read them as text files!</p>
<h4 id="definition">Definition<a href="#Definition"></a></h4>
<p>The <a href="data/ms_alignment.xsd">XML schema</a> gives a formal definition of the HSAML format.</p>
<p>An alignment generated by PRANK consist of four elements: <font color="red">newick</font>, <font color="blue">nodes</font>, <font color="green">selection</font> and <font color="orange">model</font>; sub-elements of <font color="blue">nodes</font> are either <font color="cyan">leaf</font> or <font color="purple">node</font>. Colours refer to <a href="data/hsamlexample.html">this example alignment</a>.</p>
<p>The element <font color="CC0000">newick</font> defines the alignment guide tree (or phylogeny) in the &#x27;newick&#x27; format. The tree should be complete and have a name for each node (terminal and internal) and a branch length for each branch. A node name and branch length are separated by a colon, two sister nodes separated by a comma and surrounded by a pair of brackets. A node name can be any alpha-numeric string (A-Za-z0-9 + some other characters) except that the names for internal nodes are always surrounded by hash signs (#) and the names for terminal nodes never contain hashes. It is convenient to keep the node names short and provide more information in <font color="blue">nodes</font>.</p>
<p>The element <font color="blue">nodes</font> contains terminal and internal nodes. A terminal node is defined in <font color="cyan">leaf</font> and has attributes &#x27;id&#x27;, which matches a node name in the newick tree, and &#x27;name&#x27;,the full name of the sequence. The only content of a leaf is &#x27;sequence&#x27;, an aligned sequence. An internal node is defined in <font color="purple">node</font> and has one attribute, &#x27;id&#x27; that matches a node name in the newick tree. The contents of a node are of the type &#x27;probability&#x27;, each having the attribute &#x27;id&#x27; matching one of the process id&#x27;s in the element <font color="orange">model</font>. </p>
<p>The element <font color="006600">selection</font> is optional and only appears when a user has deselected some of the alignment sites using filters based on the sitewise posterior probabilities. Selection consists of &#x27;selected_sites&#x27;, a boolean vector defining if a site is selected/deselected, and &#x27;selection_criteria&#x27;, a description of the process resulting to the current selection. </p>
<p>The element <font color="orange">model</font> has two purposes: first, it gives further information of the processes for which nodes contain posterior probabilities for (attributes &#39;id&#39; and &#39;name&#39;), and, second, it specifies how PRANKSTER displays this information. The meaning of different attributes becomes obvious if one plays with PRANKSTER (Settings-&gt;Plot), and the range of possible attribute values is defined in the <a href="data/ms_alignment.xsd">XML schema</a>.</p>
<h3 id="examples">Examples<a href="#Examples"></a></h3>
<h4 id="r">R<a href="#R"></a></h4>
<p>This R script plots the posterior probability of one of the processes across the alignment sites. If you download the <a href="data/plot_xml.rs">script here</a> and the <a href="data/CAV2.xml">example alignment here</a>, start R and type the command:</p>
<pre><code>source(&quot;plot_xml.rs&quot;)
</code></pre>
<p>you should obtain a plot similar to this:</p>
<p><img src="data/cav2.png" alt=""></p>
<p>(To reproduce the plot shown here, you should download <a href="data/plot_xml_full.rs">this script</a>.)</p>
<p>The plotted value is the posterior probability of the slow process across the alignment sites in the three internal nodes. In the two latter plots, the broad peaks match rather well with the known coding exons that are indicated by the black bars in the top (not included in the example script).</p>
<p>Note that the script requires XML package to be installed. R packages can be found from the <a href="http://www.r-project.org/">project homepage</a> following the links to CRAN and Packages. Consult the R manual to learn more about the package installation.</p>
<h4 id="perl">Perl<a href="#Perl"></a></h4>
<p>This perl script demonstrates how to select nodes and clades of an alignment tree, manipulate sequences, extract posterior probability values and work with names and id&#39;s. A detailed explanation of the use of the script and the functions implemented in the module is given in the source code file.</p>
<p>The script and the additional module required by the script can be downloaded from <a href="data/parseExample.pl">here</a> and <a href="data/Node.pm">here</a>, and the use of the script tested with this <a href="data/20prim.xml">example dataset</a>. Typing:</p>
<pre><code>perl parseExample.pl 
</code></pre>
<p>should read in the alignment and print out the location and sequence of the sites for which the alignment reliability score is below 50% in the human-chimp subtree.</p>
<p>Note that the script requires the XML::libXML module to be installed (download it from <a href="http://search.cpan.org/dist/XML-LibXML">here</a>). Consult the perl manual to learn more about the package installation.</p>
<h4 id="acknowledgements">Acknowledgements<a href="#Acknowledgements"></a></h4>
<p>The XML schema was defined by Nicolas Rodriguez.</p>
